Source/UI/BillingForm/View/BillingFormHeaderCell.swift

```diff
@@ -16,7 +16,7 @@ final class BillingFormHeaderCell: UIView {
     }()
 
     private lazy var doneButton: ButtonView = {
-        let view = ButtonView().disabledAutoresizingIntoConstraints()
+        let view = ButtonView(startEnabled: false).disabledAutoresizingIntoConstraints()
```

Comment:
please share some details why need to inject `startEnabled` f or common Button View?

Reply:
The button starts by default as enabled. Some buttons (like Pay and Billing Done) should start as disabled and only be enabled once validation passes.

This may not be visible to us in normal testing due to the speed of the action, but it becomes easily flagged by UI Test

Reply:
UI Test **must not** change normal common UI component. please undo it.

Reply:
Not sure if you read my message properly. 
The current behaviour is buggy, enabling button when it should be disabled. 
The tests did what any tests do, catch bugs.
To get the test passing, code level changes were done.

Please don't make this harder, take a minute to understand replies and ask if unclear. Don't demand things without understanding the basics behind.

Reply:
I think you misunderstand me. 
 if `ButtonView` has a new bug then it should be unit tested not a UI test.  `startEnabled ` is confusing name. you may change it to `isEnabled` like native apple button. Also, remove `isEnabled` from `ElementButtonStyle` if you want to inject button disable state.
```
public protocol ElementButtonStyle: ElementStyle {
    var isEnabled: Bool { get set }
```
Please undo changes here and a new PR to fix this issue with unit test would be great because this change will affect behaviour of buttons and needs to be tested alone.

Reply:
Ok, so different problems to address here.

 

## Testing

Unit tests and UITests are both **tests** (same as snapshots, integration etc).
The general role for testing is to find bugs, enforce behaviours and prevent unintended changes.
Different kind of tests find bugs in different ways, for example:
- **unit tests** are great at handling business logic issues, inconsistent states, parsings, data flows, etc
- **snapshot tests** help identify UI level changes and ensure with some degree of certainty the layout, positioning, formatting of the UI, etc
- **UI tests** don't do either of these. They check the usability of the UI, navigation, presence ...

> if ButtonView has a new bug then it should be unit tested not a UI test.

As I have already stated 
_Some buttons (like Pay and Billing Done) should start as disabled and only be enabled once validation passes_ 
    and 
_This may not be visible to us in normal testing due to the speed of the action, but it becomes easily flagged by UI Test_

ButtonView does not have a bug as far as Unit Tests are concerned. Furthermore ButtonView is an UIView type and Unit Testing UI components is one of the reasons MVVM exists (**to not have to**).

 

## Inject button state

The way I see these components being designed to work, they get told their configuration (via a Style interface with multiple inheritances) what their state should be.

The bug is in the way ButtonView is being enabled very early in the lifecycle of the ViewController. This is due to its default state and the delay from UI being shown to the state being updated. 

I don't think affecting that update flow is part of the solution for this problem. The solution is to be able to control the start state, which is prior to the style becoming available.

So please, if you have a preference to a different approach feel free to detail it and as always I am happy to try if technically sound.